src-tauri/src/widget_api/fs/apis.rs

```diff
@@ -0,0 +1,154 @@
+use crate::states::WidgetBaseDirectoryState;
+use std::io::Write;
+use std::path::PathBuf;
+use tauri::{command, AppHandle, Manager, Runtime};
+
+fn get_file_path<R: Runtime>(
+    app_handle: &AppHandle<R>,
+    widget_id: &str,
+    path: &str,
+) -> PathBuf {
+    let widget_base = &app_handle.state::<WidgetBaseDirectoryState>().0;
+    // if storage directory does not exist, create it
+    let widget_dir = widget_base.join(widget_id).join("storage");
+    if !widget_dir.exists() {
+        std::fs::create_dir_all(&widget_dir)
+            .expect("Failed to create widget storage directory");
+    }
+    widget_dir.join(path)
+}
+
+#[command]
+pub fn read_file<R: Runtime>(
+    app_handle: AppHandle<R>,
+    widget_id: String,
+    path: String,
+) -> Result<String, String> {
+    let file_path = get_file_path(&app_handle, &widget_id, &path);
+    std::fs::read_to_string(&file_path)
+        .map_err(|e| format!("Failed to read file '{}': {}", file_path.display(), e))
```

Comment:
Since we are using `anyhow`, I think the better way is to do:

```suggestion
    std::fs::read_to_string(&file_path).context("xxx")?;
```

This wraps an additional context `"xxx"` outside the error of `read_to_string` and propagates it. The signature of this function would then need to change. Same comment for the other commands in this file.

Reply:
Got it. Will modify all error message to use `context`.